tools/wake-format/formatter.h

```diff
@@ -347,6 +347,85 @@ struct FmtPredicate<std::initializer_list<T>> {
   bool operator()(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) { return set[node.id()]; }
 };
 
+struct EpsilonCase {
+  ALWAYS_INLINE bool run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) { return false; }
+};
+
+template <class Predicate, class FMT>
+struct PredicateCase {
+  Predicate predicate;
+  FMT formatter;
+
+  PredicateCase(Predicate predicate, FMT formatter) : predicate(predicate), formatter(formatter) {}
+
+  ALWAYS_INLINE bool run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    if (!predicate(builder, ctx, node)) {
+      return false;
+    }
+    builder.append(formatter.compose(ctx.sub(builder), node));
+    return true;
+  }
+};
+
+template <class FMT>
+struct OtherwiseCase {
+  FMT formatter;
+
+  OtherwiseCase(FMT formatter) : formatter(formatter) {}
+
+  ALWAYS_INLINE bool run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    builder.append(formatter.compose(ctx.sub(builder), node));
+    return true;
+  }
+};
+
+template <class Case1, class Case2>
+struct MatchSeq {
+  Case1 case1;
+  Case2 case2;
+  MatchSeq(Case1 c1, Case2 c2) : case1(c1), case2(c2) {}
+
+  ALWAYS_INLINE bool run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    if (case1.run(builder, ctx, node)) {
+      return true;
+    }
+    return case2.run(builder, ctx, node);
+  }
+};
+
+template <class Case>
+struct MatchAction {
+  Case c;
+
+  MatchAction(Case c) : c(c) {}
+
+  template <class FMT>
+  MatchAction<MatchSeq<Case, PredicateCase<FitsPredicate<FMT>, FMT>>> fits(FMT formatter) {
+    return {{c, {FitsPredicate<FMT>(formatter), formatter}}};
+  }
+
+  template <class FMT>
+  MatchAction<MatchSeq<Case, PredicateCase<FmtPredicate<std::initializer_list<uint8_t>>, FMT>>>
+  pred(std::initializer_list<uint8_t> ids, FMT formatter) {
+    return {{c, {ids, formatter}}};
+  }
+
+  template <class Predicate, class FMT>
+  MatchAction<MatchSeq<Case, PredicateCase<FmtPredicate<Predicate>, FMT>>> pred(Predicate predicate,
```

Comment:
Ok yeah that's the ticket, I think you *can* get the order right. You just have to use SeqFmt to do it correctly. You have all the tools I just didn't see how to do it.

I think having an `end` method that returns an AssertFmt for the default case would be good to end it.